src/Compilers/CSharp/Portable/Symbols/DelegateCaches/ModuleScopedDelegateCacheContainer.cs

```diff
@@ -0,0 +1,97 @@
+
+        private static readonly SymbolDisplayFormat MethodSortKeyFormat = new SymbolDisplayFormat
+        (
+            globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Included,
+            typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+            genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+            memberOptions: SymbolDisplayMemberOptions.IncludeType | SymbolDisplayMemberOptions.IncludeParameters,
+            parameterOptions: SymbolDisplayParameterOptions.IncludeType,
+            miscellaneousOptions: SymbolDisplayMiscellaneousOptions.EscapeKeywordIdentifiers | SymbolDisplayMiscellaneousOptions.UseSpecialTypes
+        );
+
+        private readonly TypeSymbol _delegateType;
+        private readonly FieldSymbolsCollection _delegateFields = new FieldSymbolsCollection();
+
+        private readonly Symbol _containingSymbol;
+        public override Symbol ContainingSymbol => _containingSymbol;
+
+        public override Accessibility DeclaredAccessibility => Accessibility.Internal;
+
+        public override ImmutableArray<TypeParameterSymbol> TypeParameters => ImmutableArray<TypeParameterSymbol>.Empty;
+
+        /// <remarks>This is only intended to be used from <see cref="DelegateCacheManager"/></remarks>
+        internal ModuleScopedDelegateCacheContainer(CSharpCompilation compilation, TypeSymbol delegateType)
+            : base(DelegateCacheContainerKind.ModuleScopedConcrete)
+        {
+            _containingSymbol = compilation.SourceModule.GlobalNamespace;
+            _delegateType = delegateType;
+            SortKey = delegateType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
+        }
+
+        internal override FieldSymbol ObtainCacheField(SyntheticBoundNodeFactory factory, TypeSymbol delegateType, MethodSymbol targetMethod)
+        {
+            Debug.Assert(_delegateType == delegateType);
+
+            return _delegateFields.GetOrAdd(targetMethod, m => new ModuleScopedDelegateCacheContainerField(this, m.Name, _delegateType, m.ToDisplayString(MethodSortKeyFormat)));
+        }
+
+        /// <remarks>The order should be fixed.</remarks>
+        internal ImmutableArray<ModuleScopedDelegateCacheContainerField> GetAllCreatedFields()
+        {
+            Debug.Assert(_delegateFields != null && _delegateFields.Count != 0);
+
+            var builder = ArrayBuilder<ModuleScopedDelegateCacheContainerField>.GetInstance();
+
+            foreach (var field in _delegateFields.Values)
+            {
+                builder.Add(field);
+            }
+            builder.Sort(this);
+
+            return builder.ToImmutableAndFree();
+        }
+
+        public int Compare(ModuleScopedDelegateCacheContainerField x, ModuleScopedDelegateCacheContainerField y) => String.CompareOrdinal(x.SortKey, y.SortKey);
+
+        public override ImmutableArray<Symbol> GetMembers() => StaticCast<Symbol>.From(GetAllCreatedFields());
```

Comment:
We should prevent this method from being called before all fields are added. It looks like we can use the fact that AssignName is called as an indicator that all fields have been added.